ROCmSoftwarePlatform/MIOpen
# NT_HYBRID Find mode

Experimental implementation of feature #392. This is is basically HYBRID mode that skips the most of untuned Solutions.

Smoke tested on Radeon VII, ROCm 3.5. Shows performance/behavior as expected.

### Evaluation of performance impact

Note that the tests were performed with BUILD_DEV=On and binary cache disabled. This is far from what happens when the real framework is running. However, this is more or less consistent with a situation where kernels are always compiled. I mean that the trends should be similar, i.e. if NT_HYBRID shows better wall time than HYBRID here, then I would expect the same in real life.

Find Mode | Wall time acceleration | GPU_perf_drop | Untuned Solutions disabled
-- | -- | -- | --
NORMAL | 1.00 | 0% | -
FAST | 26.44 | 51% | -
HYBRID | 1.00 | 0% | -
FAST_HYBRID | 1.55 | 0% | -
NT_HYBRID, commit (03), b8b6b9cd4 | 2.79 | 20% | All
NT_HYBRID, commit (04), f126b18b4 | 1.79 | 0% | All except ConvOclBwdWrW2<1>, ConvOclDirectFwd,   ConvOclDirectFwd1x1, ConvAsm3x3U, ConvAsmBwdWrW1x1, ConvBinWinogradRxSf2x3
NT_HYBRID, commit (05), d2ceb60bd | 2.43 | 5% | All except ConvOclBwdWrW2<1>, ConvAsm3x3U, ConvAsmBwdWrW1x1,   ConvBinWinogradRxSf2x3

### Smoke test - OK

How to smoke test:
- Clean user-find-db & run some config in NORMAL Find mode using the driver, collect logs at level 5.
  - The config should not be listed in the system find-db, for example: `./bin/MIOpenDriver conv -x 10 -y 5 -W 341 -H 79 -c 32 -n 3 -k 32 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0` (note `-n 3`)
- Clean user-find-db, run the same config in the HYBRID mode, collect log. 
- Run the same config in the same mode (do not clean find-db), collect log. 
- Clean user-find-db, run the same config in the new NT_HYBRID mode, collect log
- Run the same config in the same mode (do not clean find-db), collect log. 
 - Compare logs ;)
